File: mobile/hey-ios/Sources/HeySocial/Security/AppLock.swift
import Foundation
import LocalAuthentication

// OPTIONAL app lock — off by default. The iOS counterpart of Android's AppLock.
// When enabled, Hey asks for Face ID / Touch ID (verified in the Secure Enclave) or
// the device passcode before opening, revealing the recovery phrase, or authorizing a
// spend. This is a layer ON TOP of two things that are always true on iOS:
//   • iOS sandboxes Hey's container so other apps can't read it.
//   • iOS encrypts app data at rest (Data Protection), keyed to the device passcode.
// So the lock guards against someone holding your unlocked phone; the data itself is
// already isolated and encrypted by the OS (and additionally by SecureEnclaveVault).
enum AppLock {
    private static let key = "app_lock_enabled"
    private static var defaults: UserDefaults {
        UserDefaults(suiteName: AppPaths.appGroup) ?? .standard
    }

    /// Can this device do a biometric OR passcode check? (= Android's
    /// BIOMETRIC_STRONG | DEVICE_CREDENTIAL.)
    static func available() -> Bool {
        var err: NSError?
        return LAContext().canEvaluatePolicy(.deviceOwnerAuthentication, error: &err)
    }

    /// faceID / touchID / opticID / none — for labelling the unlock button.
    static func biometryType() -> LABiometryType {
        let ctx = LAContext()
        _ = ctx.canEvaluatePolicy(.deviceOwnerAuthentication, error: nil)
        return ctx.biometryType
    }

    static var enabled: Bool {
        get { defaults.bool(forKey: key) }
        set { defaults.set(newValue, forKey: key) }
    }

    /// Prompt for unlock. Returns true on success. `.deviceOwnerAuthentication` allows
    /// biometric OR the device passcode fallback (matches Android's allowed authenticators).
    static func prompt(reason: String = "Verify it's you to open Hey") async -> Bool {
        let ctx = LAContext()
        ctx.localizedFallbackTitle = "Use Passcode"
        guard ctx.canEvaluatePolicy(.deviceOwnerAuthentication, error: nil) else { return false }
        return await withCheckedContinuation { cont in
            ctx.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: reason) { ok, _ in
                cont.resume(returning: ok)
            }
        }
    }
}

File: mobile/hey-social-app/app/src/main/java/os/elastos/hey/social/AppLock.kt
package os.elastos.hey.social

import android.content.Context
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity

/**
 * OPTIONAL app lock — off by default. When enabled, Hey asks for the device's
 * hardware-backed biometric (fingerprint/face, verified in the TEE/StrongBox on
 * Pixel) or device PIN before opening. This is a layer ON TOP of two things that
 * are always true:
 *   • Android sandboxes Hey's storage so other apps can't read it.
 *   • Android encrypts all app data at rest (File-Based Encryption), keyed to
 *     the device lock.
 * So the lock guards against someone with your unlocked phone in hand; the data
 * itself is already isolated and encrypted by the OS.
 */
object AppLock {
    private const val PREFS = "hey"
    private const val FLAG = "app_lock"
    private const val AUTHENTICATORS =
        BiometricManager.Authenticators.BIOMETRIC_STRONG or BiometricManager.Authenticators.DEVICE_CREDENTIAL

    /** Can this device do a strong biometric or device-credential check? */
    fun available(ctx: Context): Boolean =
        BiometricManager.from(ctx).canAuthenticate(AUTHENTICATORS) == BiometricManager.BIOMETRIC_SUCCESS

    fun enabled(ctx: Context): Boolean =
        ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getBoolean(FLAG, false)

    fun setEnabled(ctx: Context, on: Boolean) =
        ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit().putBoolean(FLAG, on).apply()

    /** Prompt for unlock. [onResult] is called with true on success. */
    fun prompt(activity: FragmentActivity, onResult: (Boolean) -> Unit) {
        val prompt = BiometricPrompt(
            activity, ContextCompat.getMainExecutor(activity),
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) = onResult(true)
                override fun onAuthenticationError(code: Int, msg: CharSequence) = onResult(false)
            }
        )
        val info = BiometricPrompt.PromptInfo.Builder()
            .setTitle("Unlock Hey")
            .setSubtitle("Verify it's you to open your data")
            .setAllowedAuthenticators(AUTHENTICATORS)
            .build()
        runCatching { prompt.authenticate(info) }.onFailure { onResult(false) }
    }
}
